rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,796 @@
+                SENTENCE,
+                SEQUENCE,
+            ]:
+                continue
+            self._tf_layers[
+                f"sparse_dense.{feature_type}"
+            ] = ConcatenateSparseDenseFeatures(
+                attribute=attribute,
+                feature_type=feature_type,
+                attribute_signature=attribute_signature[feature_type],
+                config=config,
+            )
+
+        # 2. prepare layer for combining sequence- and sentence-level features
+        sentence_units = (
+            None
+            if f"sparse_dense.{SENTENCE}" not in self._tf_layers
+            else self._tf_layers[f"sparse_dense.{SENTENCE}"].output_units
+        )
+        sequence_units = (
+            None
+            if f"sparse_dense.{SEQUENCE}" not in self._tf_layers
+            else self._tf_layers[f"sparse_dense.{SEQUENCE}"].output_units
+        )
+        self._tf_layers["concat_seq_sent"] = ConcatenateSequenceSentenceFeatures(
+            attribute=attribute,
+            sequence_units=sequence_units,
+            sentence_units=sentence_units,
+            config=config,
+        )
+
+        self.output_units = self._tf_layers["concat_seq_sent"].output_units
+
+    def call(
+        self,
+        inputs: Tuple[
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            tf.Tensor,
+        ],
+        training: bool = False,
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        """Combine multiple 3-D dense/sparse feature tensors into one.
+
+        Arguments:
+            inputs: Tuple containing:
+                sequence_features: List of 3-D dense or sparse tensors with token-level
+                    features.
+                sentence_features: List of 3-D dense or sparse tensors with sentence-level
+                    features.
+                mask_sequence: a 3-D tensor mask that has 1s at real and 0s at padded
+                    positions corresponding to tokens in `sequence_features`.
+                mask_combined_sequence_sentence: a 3-D tensor mask similar to 
+                    `mask_sequence` but having each sequence of 1s longer by 1 to account
+                    for sequence lengths of sequence- and sentence-level features being
+                    combined.
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            Single dense 3-D tensor containing the combined sequence- and sentence-
+            level, sparse & dense features.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        sequence_feature_lengths = inputs[2]
+
+        mask_sequence = TransformerRasaModel._compute_mask(sequence_feature_lengths)
```

Comment:
calling private `TransformerRasaModel` doesn't feel right. If it is used so often, let's convert it into a global function

Reply:
`model_data_utils.py` I don't think is a good place. It is non tf file.

I think you can simply create global function inside `rasa_layers.py`